cognite/client/_api/datapoints.py

```diff
@@ -843,45 +819,115 @@ def custom_sort_key(x):
+        granularity,
+        include_outside_points,
+        limit,
+        ignore_unknown_ids,
+    ):
+        self.start = cognite.client.utils._time.timestamp_to_ms(start)
+        self.end = cognite.client.utils._time.timestamp_to_ms(end)
+        if len(ts_items) == 1:
+            self.aggregates = ts_items[0].get("aggregates") or aggregates
+        else:
+            self.aggregates = aggregates
+        self.ts_items = [{k: v for k, v in ts_item.items() if k in ["id", "externalId"]} for ts_item in ts_items]
+        self.granularity = granularity
+        self.include_outside_points = include_outside_points
+        self.limit = limit or float("inf")
+        self.ignore_unknown_ids = ignore_unknown_ids
+
+        self.client = client
+        self.request_limit = client._DPS_LIMIT_AGG if self.aggregates else client._DPS_LIMIT
+        self.results = defaultdict(lambda: _DPResult(self.limit))
+        self.last_result = None
+
+    def next_start_offset(self):
+        return cognite.client.utils._time.granularity_to_ms(self.granularity) if self.granularity else 1
+
+    def store_partial_result(self, request_result, start, end):
+        expected_fields = self.aggregates or ["value"]
+        if request_result:
+            result_by_identifiers = {
+                (identifier, entry[identifier]): entry
+                for entry in request_result
+                for identifier in ["id", "externalId"]
+            }
+        else:
+            result_by_identifiers = {}
+        for ts_item in self.ts_items:
+            key = next(iter(ts_item.items()))
+            result = self.results[key]
+            raw_data = result_by_identifiers.get(key)
+            if raw_data is None and self.ignore_unknown_ids:
+                result.mark_missing()
+            else:
+                result.store(
+                    raw_data, start, end, expected_fields, self.include_outside_points, self.client._cognite_client
+                )
+            self.last_result = result
+        return self.results
+
+    def as_tuples(self):
+        return [
+            (self.start, self.end, ts_item, self.aggregates, self.granularity, self.include_outside_points, self.limit)
+            for ts_item in self.ts_items
+        ]
 
 
 class DatapointsFetcher:
     def __init__(self, client: DatapointsAPI):
         self.client = client
 
     def fetch(self, query: DatapointsQuery) -> DatapointsList:
-        return self.fetch_multiple([query])[0]
+        chunk_size = self.client._TS_LIMIT
+        if query.granularity:
+            start = utils._time.timestamp_to_ms(query.start)
+            end = utils._time.timestamp_to_ms(query.end)
+            freq = utils._time.granularity_to_ms(query.granularity)
+            max_datapoint_count = int((end - start) / freq) + 1 + (2 if query.include_outside_points else 0)
+            chunk_size = min(chunk_size, self.client._DPS_LIMIT_AGG // max_datapoint_count)
+        chunk_size = max(chunk_size, 1)
```

Comment:
This looks like a good candidate for splitting into its own method with unit tests.
```suggestion
        chunk_size = self.client._TS_LIMIT
        if query.granularity:
            start = utils._time.timestamp_to_ms(query.start)
            end = utils._time.timestamp_to_ms(query.end)
            freq = utils._time.granularity_to_ms(query.granularity)
            max_datapoint_count = int((end - start) / freq) + 1 + (2 if query.include_outside_points else 0)
            chunk_size = min(chunk_size, self.client._DPS_LIMIT_AGG // max_datapoint_count)
        chunk_size = max(chunk_size, 1)
```